### PR TITLE
Fix panics with int()

### LINKED ISSUE #1228: int('0b100', base=0) panic!
## Feature

`objint.rs`:
```
/// If x is not a number or if base is given, then x must be a string,
/// bytes, or bytearray instance representing an integer literal in the
/// given base.  The literal can be preceded by '+' or '-' and be surrounded
/// by whitespace.  The base defaults to 10.  Valid bases are 0 and 2-36.
/// Base 0 means to interpret the base from the string as an integer literal.

>>> int('0b100', base=0)
```

When I run it in ubuntu 18.04, it will panic.
Not handling error correctly when out of ranges.

### PR DESCRIPTION
**Pull request by mpajkowski**
Resolves #1228 issue

### CODE REVIEW

File: tests/snippets/ints.py [hunk 1]
```diff
@@ -97,14 +97,40 @@
 assert -10 // -4 == 2
 
 assert int() == 0
-assert int("101", 2) == 5
-assert int("101", base=2) == 5
```
Comment (windelbouwman) [inline tests/snippets/ints.py]: Its good to keep this check with an explicit base given.

File: vm/src/obj/objint.rs [hunk 2]
```diff
@@ -736,21 +739,22 @@ fn int_new(cls: PyClassRef, options: IntOptions, vm: &VirtualMachine) -> PyResul
...
-                )))
+        string @ PyString => {
+            let s = string.value.as_str().trim();
+            str_to_int(vm, s, base)
+        },
+        bytes @ PyBytes => {
+            let bytes = bytes.get_value();
+            let s = std::str::from_utf8(bytes)
+                .map(|s| s.trim())
+                .map_err(|_| invalid_literal(vm, &bytes.iter().map(|&c| c as char).collect::<String>(), base))?;
```
Comment (windelbouwman) [inline vm/src/obj/objint.rs]: You might want to raise a TypeError here and use the rust error message like this:
```rust
.map_err(|e| vm.new_type_error(format!("utf8 decode error: {}", e)))?;
```

**Comment by windelbouwman** [comment]
@mpajkowski nice job!

File: vm/src/obj/objint.rs (hunk 2 above)
Comment (mpajkowski) [reply vm/src/obj/objint.rs]: I changed it to ```TypeError``` and added some tests covering that branch - but it looks that ```cpython``` uses ```ValueError``` right here.

Should we switch to ```TypeError``` or rather stay with ```cpython``` behavior?

File: tests/snippets/ints.py (hunk 1 above)
Comment (mpajkowski) [reply tests/snippets/ints.py]: ok, I'll fix it in a moment

File: vm/src/obj/objint.rs (hunk 2 above)
Comment (windelbouwman) [reply vm/src/obj/objint.rs]: Ah, okay, a `ValueError` is a better idea. I meant more the error message which is returned from the `std::from_utf8` function, we might as well use it, to indicate in a better way what went wrong.

```rust
.map_err(|e| vm.new_value_error(format!("utf8 decode error: {}", e)))?;
```

Comment (mpajkowski) [reply vm/src/obj/objint.rs]: Ok, so I'll return ```ValueError``` and apply Your suggestion :)

File: vm/src/obj/objint.rs [hunk 3]
```diff
@@ -766,6 +770,49 @@ pub fn to_int(vm: &VirtualMachine, obj: &PyObjectRef, mut base: u32) -> PyResult
     )
 }
 
+fn str_to_int(vm: &VirtualMachine, literal: &str, mut base: u32) -> PyResult<BigInt> {
+    let mut buf = literal.chars().filter(|&c| c != '_').collect::<String>();
```
Comment (seeeturtle) [inline vm/src/obj/objint.rs]: cpython has some rules for underscore in integer literal:
there can be only *one* underscore in front of the digit (not prefix)

examples:
```
>>> int('_0x121223', base=16)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid literal for int() with base 16: '_0x121223'
>>> int('121223_')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid literal for int() with base 10: '121223_'
>>> int('121__223')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid literal for int() with base 10: '121__223'
```

File: vm/src/obj/objint.rs [hunk 4]
```diff
@@ -766,6 +770,49 @@ pub fn to_int(vm: &VirtualMachine, obj: &PyObjectRef, mut base: u32) -> PyResult
...
+            } else {
+                base = matched_radix;
+            }
+            buf.drain(radix_range);
+        }
+    }
+
+    // base still not found, use default
+    if base == 0 {
+        base = 10;
```
Comment (seeeturtle) [inline vm/src/obj/objint.rs]: if base is 0 cpython doesn't allow base 10 integer starting with zero. the only acceptable case is when it is zero.
```
>>> int('0123', base=0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid literal for int() with base 0: '0123'
```
however `BigInt::from_str_radix` allows. there should be check for the case

**Pull request merged**